crates/holochain/src/core/workflow/app_validation_workflow/validation_dependencies.rs

```diff
@@ -0,0 +1,288 @@
+use std::{
+    collections::{hash_map::Entry, HashMap, HashSet},
+    time::{Duration, Instant},
+};
+
+use holo_hash::{AnyDhtHash, DhtOpHash};
+use holochain_types::dht_op::DhtOpHashed;
+
+/// In-memory struct to keep track of missing DHT hashes, which DhtOp depends on them
+/// and when they were started being fetched.
+pub struct ValidationDependencies {
+    /// Missing hashes that are being fetched, along with a set of DhtOps that depend
+    /// on the hash and the last Instant a fetch was attempted.
+    missing_hashes: HashMap<AnyDhtHash, (HashSet<DhtOpHash>, Instant)>,
+}
+
+impl Default for ValidationDependencies {
+    fn default() -> Self {
+        ValidationDependencies::new()
+    }
+}
+
+impl ValidationDependencies {
+    const FETCH_TIMEOUT: Duration = Duration::from_secs(60);
+
+    pub fn new() -> Self {
+        Self {
+            missing_hashes: HashMap::new(),
+        }
+    }
+
+    // returns true if this is a new missing hash
+    pub fn insert_missing_hash_for_op(&mut self, hash: AnyDhtHash, dht_op_hash: DhtOpHash) -> bool {
+        if let Entry::Vacant(entry) = self.missing_hashes.entry(hash.clone()) {
```

Comment:
Prefer `match` on entry because that avoids having to call `.entry` twice?

Reply:
This is verbatim what clippy told me to use instead of some `.entry().or_insert()` construct. I prefer the match too. I'll tell clippy.

Reply:
Done